### PR TITLE
Initial testing and code coverage support

### PR DESCRIPTION
**Pull request by Ramblurr**
Includes:

* test runner: mocha dependency and configuration
* code coverage: nyc/istanbul for code coverage reports
* `npm test` and `npm test:coverage`
* `gulp test` will run the tests
* if `test: true` is in the build configuration, then tests will run during the upload task
* screeps test helper to load globals
* makes game constants available at runtime
* CreepFactory using the builder pattern to help create mock creeps for tests
* one example test that works on the example codebase
* README updates

### CODE REVIEW

File: gulpfile.js
Comment (resir014) [inline gulpfile.js]: The indentation on this variable is a bit off. Be sure to use 2-space indentation.

File: test/components/creeps/creepActions.test.ts
Comment (resir014) [inline test/components/creeps/creepActions.test.ts]: Could this be done without using `/// <reference path="..." />` as per the default `tslint` rules?

**Review by resir014 — COMMENTED** [review]
I have a few changes that might have to be done based on the styleguide.

File: gulpfile.js
Comment (Ramblurr) [reply gulpfile.js]: Will do! Done. Pushed a new commit.

File: test/components/creeps/creepActions.test.ts
Comment (Ramblurr) [reply test/components/creeps/creepActions.test.ts]: Unfortunately I couldn't figure out how to do it. 

I had to add this reference path to be able to reference the global screeps typings inside the tests. 

Maybe you can take a look at it?

Comment (resir014) [reply test/components/creeps/creepActions.test.ts]: Hmm, I can't seem to figure this out. I guess we could let it stay there for now.

Just put in `/* tslint:disable:no-reference */` on the beginning of all the files that has `/// <reference path="..." />` as to not make tslint angry.

File: test/helper.ts
Comment (resir014) [inline test/helper.ts]: Might need to declare `global` to prevent any IDE errors, so do the following for this file:

```ts
declare var global: any;

import * as _ from "lodash";
import consts from "./mock/game";

global._ = _;
_.merge(global, consts);
```

File: test/lodash.ts
Comment (resir014) [inline test/lodash.ts]: Same here, do `declare var _: any;` before this line to prevent IDE errors.

**Review by resir014 — COMMENTED** [review]
I just finished a more extensive review and missed some more things.

**Comment by Ramblurr** [comment]
Alrighty, I fixed all the things you mentioned, plus quite a few others. Everything lints cleanly now, including the source dir.

**Review by resir014 — APPROVED** [review]
Cool, looks good to me. Thanks!

**Comment by Ramblurr** [comment]
Want me to squash the commits into one before merging?

**Comment by resir014** [comment]
@Ramblurr No, that's fine. I'll merge later tonight after one final test run.

**Pull request merged**
